### PR TITLE
GEODE-9191: PR clear could miss clearing bucket which lost primary

### PR DESCRIPTION
**Pull request by gesterzhou**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/internal/cache/BucketRegion.java [hunk 1]
```diff
@@ -567,19 +567,13 @@ public boolean virtualPut(EntryEventImpl event, boolean ifNew, boolean ifOld,
...
-      return;
-    }
-
     // get rvvLock
     Set<InternalDistributedMember> participants =
         getCacheDistributionAdvisor().adviseInvalidateRegion();
     boolean isLockedAlready = this.partitionedRegion.getPartitionedRegionClear()
         .isLockedForListenerAndClientNotification();
 
+    boolean lockedForPrimary = doLockForPrimary(false);
```
Comment (nabarunnag) [inline geode-core/src/main/java/org/apache/geode/internal/cache/BucketRegion.java:576]: I was just wondering how will this work if while excuting this command, the bucket is no more primary. How will this method behave in this case?

File: geode-core/src/main/java/org/apache/geode/internal/cache/InitialImageOperation.java [hunk 2]
```diff
@@ -846,12 +846,12 @@ boolean processChunk(List entries, InternalDistributedMember sender)
       }
       List<Entry> entriesToSynchronize = new ArrayList<>();
 
+      if (internalDuringApplyDelta != null && !internalDuringApplyDelta.isRunning
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/cache/InitialImageOperation.java:849]: It looks to me like changing this test hook, by moving it out of the loop, will break the existing TombstoneDUnitTest that uses DuringApplyDelta. Why did you need to change it?

File: geode-core/src/main/java/org/apache/geode/internal/cache/PartitionedRegionClear.java [hunk 3]
```diff
@@ -165,8 +165,8 @@ protected void waitForPrimary(PartitionedRegion.RetryTimeKeeper retryTimer) {
                     new RegionEventImpl(localPrimaryBucketRegion, Operation.REGION_CLEAR, null,
                         false, partitionedRegion.getMyId(), regionEvent.getEventId());
                 localPrimaryBucketRegion.cmnClearRegion(bucketRegionEvent, false, true);
+                clearedBuckets.add(localPrimaryBucketRegion.getId());
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/cache/PartitionedRegionClear.java:168]: How is clearedBuckets used? You are changing the code to only add a bucket to it if the size of the bucket is > 0. I can see how clearedBuckets is used down at line 184 to decide if we will increment stats. This does not seem right. If we find a bunch of empty bucket then we still did a clear and took some time so it seems wrong to not even change the stats.
Also we return clearedBuckets from this method and I can't tell what the callers do with it.
Also this check that the primary bucket is not empty does not seem safe to me. Currently if evict destroy is configured on a PR then the primary and secondary can be out of sync. So the primary could be empty but the secondary could still have entries.

File: geode-core/src/main/java/org/apache/geode/internal/cache/BucketRegion.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (dschneider-pivotal) [reply geode-core/src/main/java/org/apache/geode/internal/cache/BucketRegion.java:576]: Perhaps what you really want this method to do on a secondary is to just call clearRegionLocally(regionEvent, cacheWrite, null); All the locking and distribution stuff has been done on the primary

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIRequesterDUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+          }
+
+          @Override
+          public String description() {
+            return null;
+          }
+        };
+
+        GeodeAwaitility.await().untilAsserted(ev);
+        if (callback == null || !callback.isRunning) {
```
Comment (DonalEvans) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIRequesterDUnitTest.java:288]: It should not be necessary to check these conditions here or throw if they're not met, as the `await()` call above will throw if it times out waiting for the callback to be running.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIRequesterDUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+          fail("GII tesk hook is not started yet");
+        }
+      }
+    };
+    vm.invoke(waitForCallbackStarted);
+  }
+
+  private static class PauseDuringGIICallback extends InitialImageOperation.GIITestHook {
+    private Object lockObject = new Object();
```
Comment (DonalEvans) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIRequesterDUnitTest.java:297]: A couple of warnings can be cleared up if this object is made `final`

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIRequesterDUnitTest.java [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+
+  public void waitForGIITeskHookStarted(final MemberVM vm,
+      final InitialImageOperation.GIITestHookType callbacktype) {
+    SerializableRunnable waitForCallbackStarted = new SerializableRunnable() {
+      @Override
+      public void run() {
+
+        final InitialImageOperation.GIITestHook callback =
+            getGIITestHookForCheckingPurpose(callbacktype);
+        WaitCriterion ev = new WaitCriterion() {
```
Comment (DonalEvans) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIRequesterDUnitTest.java:274]: `WaitCriterion` is deprecated and should not be used. Instead, `GeodeAwaitility.await()` should be used to assert the conditions, e.g. 
```
GeodeAwaitility.await().untilAsserted(() -> assertThat(getGIITestHookForCheckingPurpose(callbacktype)).satisfies(callback -> {
          assertThat(callback).isNotNull();
          assertThat(callback.isRunning).isTrue();
        }));
```

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIRequesterDUnitTest.java [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+      }
+    }
+
+    @Override
+    public void run() {
+      synchronized (this.lockObject) {
+        try {
+          isRunning = true;
+          this.lockObject.wait();
+        } catch (InterruptedException e) {
```
Comment (DonalEvans) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIRequesterDUnitTest.java:316]: IntelliJ warning here can be fixed if the exception name is changed to "ignore".

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIRequesterDUnitTest.java [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+        }
+      }
+    }
+  } // Mycallback
+
+  private class PauseDuringClearDistributionMessageObserver
+      extends DistributionMessageObserver {
+    private CountDownLatch latch = new CountDownLatch(1);
```
Comment (DonalEvans) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIRequesterDUnitTest.java:324]: This can be `final`

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIRequesterDUnitTest.java [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+
+    @Override
+    public void beforeProcessMessage(ClusterDistributionManager dm, DistributionMessage message) {
+      if (message instanceof PartitionedRegionClearMessage) {
+        PartitionedRegionClearMessage prcm = (PartitionedRegionClearMessage) message;
```
Comment (DonalEvans) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIRequesterDUnitTest.java:329]: The variable `prcm` is never used, so this line can be removed.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIRequesterDUnitTest.java [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
+        try {
+          logger.info("before wait for clear message");
+          latch.await();
+          logger.info("after wait for clear message");
+        } catch (InterruptedException ex) {
```
Comment (DonalEvans) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIRequesterDUnitTest.java:334]: IntelliJ warning here can be fixed by renaming this exception "ignore".

**Review by DonalEvans — CHANGES_REQUESTED** [review]
Every comment in `ClearPRDuringGIIRequesterDUnitTest.java` also applies to `ClearPRDuringGIIProviderDUnitTest.java` since they're very similar classes.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIRequesterDUnitTest.java [hunk 11]
```diff
@@ -0,0 +1,339 @@
...
+    }
+  }
+
+  @Test
+  public void clearBeforeGIIShouldClearTheRegion() {
+    giiTestHookSyncWithClear(true);
+  }
+
+  @Test
+  public void clearAfterGIIShouldClearTheRegion() {
```
[the remark this answers is not in the document]
Comment (gesterzhou) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIRequesterDUnitTest.java:247]: PartitionedRegionPartialClearException is an expected behavior, which will be thrown to the user.

File: geode-core/src/main/java/org/apache/geode/internal/cache/InitialImageOperation.java (hunk 2 above)
Comment (gesterzhou) [reply geode-core/src/main/java/org/apache/geode/internal/cache/InitialImageOperation.java:849]: because I need to make sure processChunk is called. If I did not move out of the loop, if the entryCount is 0, the testHook will not be called. I don't want to introduce another testHook. 

I care if processChunk is call more than if there's any data in chunk.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIRequesterDUnitTest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (DonalEvans) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIRequesterDUnitTest.java:247]: If the PartitionedRegionPartialClearException is expected, then the test should not be asserting that the region is cleared, it should only assert that a PartitionedRegionPartialClearException is thrown, and the test name should be changed to reflect that.

File: geode-core/src/main/java/org/apache/geode/internal/cache/PartitionedRegionClear.java (hunk 3 above)
Comment (gesterzhou) [reply geode-core/src/main/java/org/apache/geode/internal/cache/PartitionedRegionClear.java:168]: We definitely should remove the "if (localPrimaryBucketRegion.size() > 0) {"

File: geode-core/src/main/java/org/apache/geode/internal/cache/PartitionedRegionClearMessage.java [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+import org.apache.geode.internal.CopyOnWriteHashSet;
+import org.apache.geode.internal.NanoTimer;
+import org.apache.geode.internal.cache.partitioned.PartitionMessage;
+import org.apache.geode.internal.logging.log4j.LogMarker;
+import org.apache.geode.internal.serialization.DeserializationContext;
+import org.apache.geode.internal.serialization.SerializationContext;
+import org.apache.geode.logging.internal.log4j.api.LogService;
+
+/**
+ * this message is for operations no the partition region level, could be sent by any originating
```
Comment (Bill) [inline geode-core/src/main/java/org/apache/geode/internal/cache/PartitionedRegionClearMessage.java:45]: typo replace "no" with "on"

**Review by rhoughton-pivot — APPROVED** [review]

File: geode-core/src/main/java/org/apache/geode/internal/cache/partitioned/ClearPRMessage.java [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+import org.apache.geode.internal.cache.EventID;
+import org.apache.geode.internal.cache.ForceReattemptException;
+import org.apache.geode.internal.cache.PartitionedRegion;
+import org.apache.geode.internal.cache.RegionEventImpl;
+import org.apache.geode.internal.logging.log4j.LogMarker;
+import org.apache.geode.internal.serialization.DeserializationContext;
+import org.apache.geode.internal.serialization.SerializationContext;
+import org.apache.geode.logging.internal.log4j.api.LogService;
+
+public class ClearPRMessage extends PartitionMessageWithDirectReply {
```
Comment (Bill) [inline geode-core/src/main/java/org/apache/geode/internal/cache/partitioned/ClearPRMessage.java:54]: Need a class-level documentation comment explaining what this class is for.

This is especially needed because this PR introduces another, very similarly-named message class `PartitionedRegionClearMessage`. I have no idea why both are needed and how they relate.

**Review by Bill — CHANGES_REQUESTED** [review]
In addition to the changes requested inline, I'd like to ask if you could make the naming a bit more consistent. I see these new names:

```
CLEAR_PARTITIONED_REGION_MESSAGE
PartitionedRegionClearMessage

CLEAR_PARTITIONED_REGION_REPLY_MESSAGE
PartitionedRegionClearMessage.PartitionedRegionClearReplyMessage

PR_CLEAR_MESSAGE
ClearPRMessage

PR_CLEAR_REPLY_MESSAGE
ClearPRMessage.ClearReplyMessage
```

In some cases "partitioned region" is abbreviated. In other cases it is not.

In some cases "partitioned region" (or its abbreviation) comes first. In other cases it does not.

Making the constant names match their class names better, and making all the names follow a consistent pattern will help maintainers know what to expect.

**Comment by jinmeiliao** [comment]
Is it possible to break this PR down to multiple PRs. The changes seem overwhelming. Let's separate functionality change from refactoring (renaming etc.)

**Comment by DonalEvans** [comment]
> Is it possible to break this PR down to multiple PRs. The changes seem overwhelming. Let's separate functionality change from refactoring (renaming etc.)

The reason the PR is so huge is that the GEODE-7665 branch was rebased on develop after this PR was created. The PR needs to be rebased on the GEODE-7665 branch before it can be reviewed. @gesterzhou

File: geode-core/src/distributedTest/java/org/apache/geode/cache/query/partitioned/PRClearCreateIndexDUnitTest.java [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+
+  // Note: OP_LOCK_FOR_CLEAR, OP_CLEAR, OP_UNLOCK_FOR_CLEAR are messages for secondary members
+  // OP_LOCK_FOR_PR_CLEAR, OP_UNLOCK_FOR_PR_CLEAR, OP_PR_CLEAR can be for anybody
+
+  @Test
+  // all local buckets are primary, so only OP_LOCK_FOR_CLEAR and OP_CLEAR messages are sent to the
+  // secondary member
+  // in the end an OP_PR_CLEAR is sent to the secondary for no effect
+  public void clearFromPrimaryMember() throws Exception {
+    AsyncInvocation createIndex = secondary.invokeAsync(PRClearCreateIndexDUnitTest::createIndex);
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/cache/query/partitioned/PRClearCreateIndexDUnitTest.java:98]: When there's no return from an invokeAsync, the type should be `AsyncInvocation<Void>`. In general, we should always declare types on everything.

File: geode-core/src/distributedTest/java/org/apache/geode/cache/query/partitioned/PRClearCreateIndexDUnitTest.java [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+    AsyncInvocation clear = primary.invokeAsync(PRClearCreateIndexDUnitTest::clear);
+
+    createIndex.get();
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/cache/query/partitioned/PRClearCreateIndexDUnitTest.java:101]: These calls to `AsyncInvocation.get()` should actually be `AsyncInvocation.await()` since there's no return

File: geode-core/src/distributedTest/java/org/apache/geode/cache/query/partitioned/PRClearCreateIndexDUnitTest.java [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+import org.apache.geode.test.dunit.AsyncInvocation;
+import org.apache.geode.test.dunit.rules.ClientVM;
+import org.apache.geode.test.dunit.rules.ClusterStartupRule;
+import org.apache.geode.test.dunit.rules.MemberVM;
+
+public class PRClearCreateIndexDUnitTest implements Serializable {
+  @Rule
+  public ClusterStartupRule cluster = new ClusterStartupRule(4, true);
+
+  private MemberVM primary, secondary;
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/cache/query/partitioned/PRClearCreateIndexDUnitTest.java:47]: The spotless format really shouldn't be allowing this. Please declare each field on its own line.

File: geode-core/src/distributedTest/java/org/apache/geode/cache/query/partitioned/PRClearQueryIndexDUnitTest.java [hunk 17]
```diff
@@ -0,0 +1,376 @@
...
+  }
+
+  @Test
+  public void concurrentClearAndPut() throws Exception {
+    AsyncInvocation puts = server1.invokeAsync(() -> {
+      Cache cache = ClusterStartupRule.getCache();
+      Region region = cache.getRegion("cities");
+      for (int i = 0; i < 1000; i++) {
+        // wait for gate to open
+        getBlackboard().waitForGate("proceedToPut", 60, TimeUnit.SECONDS);
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/cache/query/partitioned/PRClearQueryIndexDUnitTest.java:298]: We shouldn't be using timeouts less than the default that is in GeodeAwaitility. You can simply use:
```
getBlackboard().waitForGate("proceedToPut", GeodeAwaitility.getTimeout().getSeconds(), TimeUnit.SECONDS);
```

File: geode-core/src/distributedTest/java/org/apache/geode/cache/query/partitioned/PRClearQueryIndexDUnitTest.java [hunk 18]
```diff
@@ -0,0 +1,376 @@
...
+    startStopServer.get(60, TimeUnit.SECONDS);
+    putAndClear.get(60, TimeUnit.SECONDS);
+  }
+
+  private static DUnitBlackboard getBlackboard() {
+    if (blackboard == null) {
+      blackboard = new DUnitBlackboard();
+    }
+    return blackboard;
+  }
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/cache/query/partitioned/PRClearQueryIndexDUnitTest.java:364]: Please use `DistributedBlackboard` instead:
```
  @Rule
  public DistributedBlackboard blackboard = new DistributedBlackboard();
```

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIProviderDUnitTest.java [hunk 19]
```diff
@@ -0,0 +1,325 @@
...
+
+  public void waitForGIITeskHookStarted(final MemberVM vm,
+      final InitialImageOperation.GIITestHookType callbacktype) {
+    SerializableRunnable waitForCallbackStarted = new SerializableRunnable() {
+      @Override
+      public void run() {
+
+        final InitialImageOperation.GIITestHook callback =
+            getGIITestHookForCheckingPurpose(callbacktype);
+        WaitCriterion ev = new WaitCriterion() {
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIProviderDUnitTest.java:260]: Please don't use `WaitCriterion` anymore. You can just directly specify the content of `done()` within the call to `await()`:
```
GeodeAwaitility.await().until(() -> callback != null && callback.isRunning);
```
Note that you want to use `await().until` instead of `await().untilAsserted` since `callback != null && callback.isRunning` never throws anything.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIProviderDUnitTest.java [hunk 20]
```diff
@@ -0,0 +1,325 @@
...
+          @Override
+          public String description() {
+            return null;
+          }
+        };
+
+        GeodeAwaitility.await().untilAsserted(ev);
+        if (callback == null || !callback.isRunning) {
+          fail("GII tesk hook is not started yet");
+        }
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIProviderDUnitTest.java:276]: Please use an assertions instead:
```
assertThat(callback).isNotNull();
assertThat(callback.isRunning).isTrue();
```
In general, I would try to avoid having something like `callback` be null so that you don't need the `isNotNull`.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIProviderDUnitTest.java [hunk 21]
```diff
@@ -0,0 +1,325 @@
...
+    }
+
+    @Override
+    public void run() {
+      synchronized (this.lockObject) {
+        try {
+          isRunning = true;
+          this.lockObject.wait();
+        } catch (InterruptedException e) {
+        }
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIProviderDUnitTest.java:303]: Please use ErrorCollector or DistributedErrorCollector instead of eating exceptions:
```
  @Rule
  public DistributedErrorCollector errorCollector = new DistributedErrorCollector();
```
```
} catch (InterruptedException e) {
  errorCollector.addError(e);
}
```
You'll need to change PauseDuringGIICallback to non-static to use the Rule or change PauseDuringGIICallback to have a field of type DistributedErrorCollector which is passed in.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIProviderDUnitTest.java [hunk 22]
```diff
@@ -0,0 +1,325 @@
...
+        if (callback == null || !callback.isRunning) {
+          fail("GII tesk hook is not started yet");
+        }
+      }
+    };
+    vm.invoke(waitForCallbackStarted);
+  }
+
+  private static class PauseDuringGIICallback extends InitialImageOperation.GIITestHook {
+    private Object lockObject = new Object();
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIProviderDUnitTest.java:283]: Any synchronization object should be final, so just add the final keyword.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIProviderDUnitTest.java [hunk 23]
```diff
@@ -0,0 +1,325 @@
...
+          this.lockObject.wait();
+        } catch (InterruptedException e) {
+        }
+      }
+    }
+  } // Mycallback
+
+  private class PauseDuringClearDistributionMessageObserver
+      extends DistributionMessageObserver {
+    private CountDownLatch latch = new CountDownLatch(1);
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClearPRDuringGIIProviderDUnitTest.java:310]: You'll want to have a tearDown in the test class that counts this down. This will prevent the test from leaving an orphaned thread.

In general, it's cleaner to have the test own the CountDownLatch and pass a reference into this inner class via a constructor.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionAfterClearNotificationDUnitTest.java [hunk 24]
```diff
@@ -0,0 +1,372 @@
...
+  protected MemberVM locator;
+  protected MemberVM dataStore1;
+  protected MemberVM dataStore2;
+  protected MemberVM dataStore3;
+  protected MemberVM accessor;
+
+  protected ClientVM client1;
+  protected ClientVM client2;
+
+  private static volatile DUnitBlackboard blackboard;
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionAfterClearNotificationDUnitTest.java:66]: Please use the DistributedBlackboard rule instead:
```
  @Rule
  public DistributedBlackboard blackboard = new DistributedBlackboard();
```

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionAfterClearNotificationDUnitTest.java [hunk 25]
```diff
@@ -0,0 +1,372 @@
...
+    accessor.invoke(() -> feed(false));
+    verifyServerRegionSize(NUM_ENTRIES);
+
+    dataStore1.invoke(() -> getRegion(false).clear());
+
+    verifyServerRegionSize(0);
+    verifyCacheListenerTriggerCount(dataStore1);
+  }
+
+  @Test(expected = AssertionError.class)
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionAfterClearNotificationDUnitTest.java:250]: Please use AssertJ's `catchThrowable` instead. If any line throws AssertionError, this test will still fail.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionAfterClearNotificationDUnitTest.java [hunk 26]
```diff
@@ -0,0 +1,372 @@
...
+    client2.invoke(this::initClientCache);
+
+    accessor.invoke(() -> feed(false));
+    verifyServerRegionSize(NUM_ENTRIES);
+    dataStore2.invoke(() -> DistributionMessageObserver.setInstance(
+        testHookToKillMemberCallingClearBeforeMessageProcessed()));
+
+    AsyncInvocation ds1ClearAsync = dataStore1.invokeAsync(() -> getRegion(false).clear());
+
+    getBlackboard().waitForGate("CLOSE_CACHE", 30, SECONDS);
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionAfterClearNotificationDUnitTest.java:263]: All of these timeouts should change to `GeodeAwaitility.getTimeout().getSeconds()` to avoid intermittent failures on slow machines.

**Review by kirklund — CHANGES_REQUESTED** [review]
